### PR TITLE
Add support eMIOS PWM pulse generation for S32K344

### PR DESCRIPTION
**Pull request by Dat-NguyenDuy**
This introduces PWM shim driver for NXP S32 eMIOS with APIs for pulse generation

### CODE REVIEW

**Comment by manuargue** [comment]
hi @Dat-NguyenDuy pls fix the failed compliance checks and builds

**Comment by Dat-NguyenDuy** [comment]
> hi @Dat-NguyenDuy pls fix the failed compliance checks and builds

Hi, fixed the compliance

File: dts/bindings/pwm/nxp,s32-emios-pwm.yaml [hunk 1]
```diff
@@ -0,0 +1,128 @@
+description: |
+  NXP S32 eMIOS PWM node for S32K3 SoCs. Each channel in eMIOS can be configured
```
Comment (manuargue) [inline dts/bindings/pwm/nxp,s32-emios-pwm.yaml:2]: does this binding only applies for S32K3? I'd expect it can be reused in other SoCs like S32Z

File: modules/Kconfig.s32 [hunk 2]
```diff
@@ -1,9 +1,13 @@
...
-# Copyright 2022 NXP
+# Copyright 2022-2023 NXP
 # SPDX-License-Identifier: Apache-2.0
 
 config HAS_S32_HAL
 	bool
 	select HAS_CMSIS_CORE
 	depends on SOC_FAMILY_S32
+
+config NOCACHE_LOAD_MEMORY
```
Comment (manuargue) [inline modules/Kconfig.s32:11]: perhaps we could consider making this available to other SoCs? similarly to `CONFIG_NOCACHE_MEMORY`

File: dts/bindings/timer/nxp,s32-emios-timer.yaml [hunk 3]
```diff
@@ -0,0 +1,94 @@
+# Copyright 2023 NXP
```
Comment (mmahadevan108) [inline dts/bindings/timer/nxp,s32-emios-timer.yaml:1]: Why is this file added. I don't see a timer driver. If this module is intended for a timer driver then this binding should be included as part of that PR.

Comment (Dat-NguyenDuy) [reply dts/bindings/timer/nxp,s32-emios-timer.yaml:1]: No, this module is not intended for timer driver (although it could), i will push the update to move it out of here

File: dts/bindings/pwm/nxp,s32-emios-pwm.yaml (hunk 1 above)
Comment (Dat-NguyenDuy) [reply dts/bindings/pwm/nxp,s32-emios-pwm.yaml]: Right now is fit for S32K3 but yes can be extended, updated

File: drivers/pwm/pwm_nxp_s32_emios.c [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+#define BUS_C				EMIOS_PWM_IP_BUS_BCDE
+#define BUS_D				EMIOS_PWM_IP_BUS_BCDE
+#define BUS_E				EMIOS_PWM_IP_BUS_BCDE
+#define BUS_F				EMIOS_PWM_IP_BUS_F
+
+#define EMIOS_MASTER_BUS_MODE(mode)	DT_CAT(EMIOS_IP_, mode)
+#define EMIOS_PWM_MODE(mode)		DT_CAT(EMIOS_PWM_IP_, mode)
+#define EMIOS_PWM_MODE_WITH_FLAG(mode)	DT_CAT3(EMIOS_PWM_IP_, mode, _FLAG)
+#define EMIOS_PWM_POLARITY(mode)	DT_CAT(EMIOS_PWM_IP_, mode)
+#define EMIOS_PWM_PRESCALER_SRC(mode)	DT_CAT(EMIOS_PWM_IP_PS_SRC_, mode)
```
Comment (danieldegrasse) [inline drivers/pwm/pwm_nxp_s32_emios.c:32]: Nit- can we move these macros to the bottom of the file, where they are used?

File: modules/Kconfig.s32 (hunk 2 above)
Comment (mmahadevan108) [reply modules/Kconfig.s32:11]: Can you add a description for this Kconfig

Comment (Dat-NguyenDuy) [reply modules/Kconfig.s32]: When i revisited the bare-metal driver and had a discussion with the engineer from HAL team, he said that actually the variable is supposed and can be put in to a read only region. We will revisit the whole supporting features and if needed, will put a separate PR for this support

File: dts/bindings/timer/nxp,s32-emios-timer.yaml (hunk 3 above)
Comment (Dat-NguyenDuy) [reply dts/bindings/timer/nxp,s32-emios-timer.yaml:1]: I created separate misc driver for eMIOS that does initializes global configuration for eMIOS

File: drivers/pwm/pwm_nxp_s32_emios.c (hunk 4 above)
Comment (Dat-NguyenDuy) [reply drivers/pwm/pwm_nxp_s32_emios.c]: I cleaned up a little bit and move to bottom of the file

File: dts/bindings/timer/nxp,s32-emios-timer.yaml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  NXP S32 Enhanced Modular IO SubSystem (eMIOS) node for S32K3 SoCs.
```
[the remark this answers is not in the document]
Comment (Dat-NguyenDuy) [reply dts/bindings/timer/nxp,s32-emios-timer.yaml]: Right now is fit for S32K3 but yes can be extended, updated

**Review by danieldegrasse — DISMISSED** [review]

File: drivers/misc/nxp_s32_emios/nxp_s32_emios.c [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+#define NXP_S32_EMIOS_MASTER_BUS_MODE(mode)	DT_CAT(EMIOS_IP_, mode)
+
+#define NXP_S32_EMIOS_INSTANCE_CHECK(idx, n)							\
+	((DT_INST_REG_ADDR(n) == IP_EMIOS_##idx##_BASE) ? idx : 0)
+
+#define NXP_S32_EMIOS_GET_INSTANCE(n)								\
+	LISTIFY(__DEBRACKET eMIOS_INSTANCE_COUNT, NXP_S32_EMIOS_INSTANCE_CHECK, (|), n)
+
+/* Consider add comment for GlobalTimeBase */
```
Comment (manuargue) [inline drivers/misc/nxp_s32_emios/nxp_s32_emios.c:45]: nit: should this comment be removed?

File: drivers/misc/nxp_s32_emios/nxp_s32_emios.c [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+#define NXP_S32_EMIOS_GENERATE_GLOBAL_CONFIG(n)							\
+	BUILD_ASSERT(DT_INST_PROP(n, clock_divider) <= 0xFFU,					\
```
Comment (manuargue) [inline drivers/misc/nxp_s32_emios/nxp_s32_emios.c:47]: Consider adding a macro definition for `0xff`

**Comment by mmahadevan108** [comment]
@Dat-NguyenDuy , please update `west.yml`

File: dts/bindings/misc/nxp,s32-emios.yaml [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+properties:
+  reg:
+    required: true
+
+  clock-divider:
+    type: int
+    required: true
+    description: |
+      Clock divider value for the global prescaler. Could be in range [1 .. 256]
```
Comment (manuargue) [inline dts/bindings/misc/nxp,s32-emios.yaml:22]: Isn't the build assertion in the shim driver checking that this value is between 0 and 255?

File: dts/bindings/misc/nxp,s32-emios.yaml [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+child-binding:
+  child-binding:
+    description: |
+      Node for eMIOS master bus. Each channel is capable to become a master bus has
+      a node defined in root devicetree but is disabled by default. To allow using
+      the master bus, the devicetree node should be enabled and dts properties
+      should be configured as required by application. For example, to enable bus A
+      of eMIOS instance 0:
+        master_bus {
+          emios0_bus_a: emios0_bus_a {
```
Comment (manuargue) [inline dts/bindings/misc/nxp,s32-emios.yaml:33]: There are more properties required than the ones included in this snippet. Perhaps add them all to have a complete example.

**Review by manuargue — COMMENTED** [review]
Some minor nits left from my side

File: dts/bindings/misc/nxp,s32-emios.yaml (hunk 9 above)
Comment (Dat-NguyenDuy) [reply dts/bindings/misc/nxp,s32-emios.yaml:35]: Yep, i added for more details

File: dts/bindings/misc/nxp,s32-emios.yaml (hunk 8 above)
Comment (Dat-NguyenDuy) [reply dts/bindings/misc/nxp,s32-emios.yaml]: Right, add check for range [1 256]

File: drivers/misc/nxp_s32_emios/nxp_s32_emios.c (hunk 6 above)
Comment (Dat-NguyenDuy) [reply drivers/misc/nxp_s32_emios/nxp_s32_emios.c]: I removed

**Review by manuargue — DISMISSED** [review]
All my comments addressed. Failed builds seems to be caused by other boards unrelated to these changes.

**Comment by Dat-NguyenDuy** [comment]
For imxrt, samples\drivers\led_pwm is build failing look like because of this updates https://github.com/zephyrproject-rtos/hal_nxp/pull/259. The test still passes on main because that one was merged in hal_nx[ but somehow the manifest in main was not updated accordingly

**Comment by Dat-NguyenDuy** [comment]
Rebase main to pick up https://github.com/zephyrproject-rtos/zephyr/pull/62050, that fixed the issue causes some nrf boards build failure

**Comment by mmahadevan108** [comment]
Could you please rebase. The CI failure related to `RT 1170` has been addressed and merged.

**Comment by Dat-NguyenDuy** [comment]
> Could you please rebase. The CI failure related to `RT 1170` has been addressed and merged.

Thanks, i rebased

**Review by manuargue — APPROVED** [review]
All my comments address, approved.

@bperseghetti if you can spend some time reviewing these changes it would be valuable to get feedback from you, since you are working on the integration of this board in a larger system.

**Review by bperseghetti — APPROVED** [review]

**Comment by bperseghetti** [comment]
> All my comments address, approved.
> 
> @bperseghetti if you can spend some time reviewing these changes it would be valuable to get feedback from you, since you are working on the integration of this board in a larger system.

I think overall it looks good and is a much needed add for the board/SoC. I think my only things are naming nits here and there but those are just preferential, I don't see any issues from reviewing over it.
